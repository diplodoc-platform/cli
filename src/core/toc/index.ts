export type * from './types';
export type {LoaderContext} from './loader';

export {IncludeMode} from './loader';
export {TocService} from './TocService';
export {GenericIncluderExtension} from './includers/generic';
export {OpenapiIncluderExtension} from './includers/openapi';
