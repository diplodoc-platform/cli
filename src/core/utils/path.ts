import {dirname, isAbsolute, join, normalize} from 'node:path';
import _normalizePath from 'normalize-path';

import {isExternalHref} from '~/core/utils/url';

export function normalizePath(path: string): NormalizedPath {
    return _normalizePath(normalize(_normalizePath(path, false)), false);
}

export function isRelativePath(path: string): path is RelativePath {
    return !isExternalHref(path) && !isAbsolute(path);
}

export function langFromPath(path: string, config: {lang?: string; langs: string[]}) {
    const {lang, langs} = config;
    const pathBaseLang = normalizePath(path).split('/')[0];
    const pathLang = langs.includes(pathBaseLang) && pathBaseLang;

    return pathLang || lang || langs[0];
}

export function rebasePath(root: RelativePath, path: RelativePath) {
    return normalizePath(join(dirname(root), path));
}

export function fullPath(
    path: AbsolutePath | NormalizedPath,
    root: NormalizedPath,
): NormalizedPath {
    if (path.match(/^(\/|\\)/)) {
        return normalizePath(path.slice(1));
    } else {
        return normalizePath(join(dirname(root), path));
    }
}
