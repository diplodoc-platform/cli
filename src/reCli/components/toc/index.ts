export * from './toc';
export * from './transform';
