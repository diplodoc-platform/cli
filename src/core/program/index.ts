import type {BaseArgs, BaseConfig, ExtensionInfo, ICallable} from './types';
import type {Command, Config, ExtendedOption} from '~/core/config';
import Module from 'node:module';
import {dirname, isAbsolute, resolve} from 'node:path';
import {once, pick} from 'lodash';

import {
    resolveConfig,
    scope as scopeConfig,
    strictScope as strictScopeConfig,
    withConfigUtils,
} from '~/core/config';
import {Logger, stats} from '~/core/logger';

import {getHooks, withHooks} from './hooks';
import {HandledError} from './utils';
import {getConfigDefaults, getConfigScope, withConfigDefaults, withConfigScope} from './decorators';

export * from './types';

export {getHooks, withConfigDefaults, withConfigScope};

const isRelative = (path: string | undefined) => /^\.{1,2}\//.test(path || '');

const YFM_CONFIG_FILENAME = '.yfm';

type Behavior = {
    isDefaultCommand?: boolean;
};

@withHooks
@withConfigDefaults(() => ({
    strict: false,
    quiet: false,
}))
/**
 * Program should follow some simple rules:
 * 1. It is a base independent unit which can contain subprograms or do something by itself.
 *    (NOT BOTH)
 * 2. Required 'apply' method serves for: ```
 *   - initial data binding
 *   - subprograms initialisation
 *   - hooks subscription
 *   - error handling
 * ```
 * 3. Program can be subprogram. This can be detected by non empty param passed to `apply` method.
 *    But anyway program should be independent unit.
 * 4. In most cases hook **execution** in 'apply' will be architecture mistake.
 * 5. Optional 'action' method - is a main place for hooks **execution**.
 *    For compatibility with Commander.Command->action method result should be void.
 * 6. Complex hook calls should be designed as external private methods named as 'hookMethodName'
 *    (example: hookConfig)
 */
export class BaseProgram<
    TConfig extends BaseConfig = BaseConfig,
    TArgs extends BaseArgs = BaseArgs,
> {
    static is(program: BaseProgram) {
        return program instanceof this;
    }

    readonly name: string = 'Base';

    readonly command!: Command;

    readonly config!: Config<TConfig>;

    readonly logger: Logger = new Logger();

    readonly options!: ExtendedOption[];

    protected modules: ICallable[] = [];

    protected extensions: (string | ExtensionInfo)[] = [];

    private parent: BaseProgram | null = null;

    private behavior: Behavior = {};

    constructor(config?: BaseConfig & TConfig, behavior: Behavior = {}) {
        this.behavior = behavior;

        if (config) {
            const defaults = getConfigDefaults(this);
            this.config = withConfigUtils(process.cwd(), {
                ...defaults,
                ...config,
            });
        }
    }

    async init(args: BaseArgs, parent?: BaseProgram, isExtension = false) {
        this.logger.setup(args);

        const config = await this.resolveConfig(args as TArgs);
        const extensions = await this.resolveExtensions(config, args);

        // @ts-ignore
        this['config'] = parent?.config || (await this.hookConfig(config, args as TArgs));

        this.logger.setup(this['config']);

        this.modules.push(...extensions);

        this.options.forEach((option) => {
            this.command.addOption(option);
        });

        this.command.action(() => this._action());

        for (const module of this.modules) {
            if (isProgram(module)) {
                await module.init(args, this);
            } else {
                module.apply(this);
            }
        }

        this.apply(parent);
    }

    apply(program?: BaseProgram) {
        const isDefault = this.behavior.isDefaultCommand;
        // @ts-ignore
        this['parent'] = program;

        if (program) {
            getHooks(program).Command.tap(
                this.name,
                once((command) => {
                    command.addCommand(this.command, {isDefault});
                    getHooks(this).Command.call(this.command, this.options);
                }),
            );
            this.logger.pipe(program.logger);
        } else {
            getHooks(this).Command.call(this.command, this.options);
        }
    }

    async parse(args: string[]) {
        return this.command.parseAsync(args);
    }

    async action(_args: TArgs) {
        throw new Error('Should be implemented');
    }

    args(args: Hash): TArgs {
        const options = this.options.map((option) => option.attributeName());

        return {
            ...this.parent?.args(args),
            ...pick(args, options),
        } as TArgs;
    }

    private async resolveConfig(args: TArgs) {
        const defaults = getConfigDefaults(this);
        const {scope, strictScope} = getConfigScope(this);
        const configPath =
            isAbsolute(args.config) || isRelative(args.config)
                ? resolve(args.config)
                : resolve(args.input, args.config);

        const filter =
            (strictScope && strictScopeConfig(strictScope)) || (scope && scopeConfig(scope));

        return resolveConfig(configPath, {
            filter: filter || undefined,
            defaults: defaults,
            fallback: args.config === YFM_CONFIG_FILENAME ? defaults : null,
        });
    }

    private async hookConfig(config: Config<TConfig>, args: TArgs) {
        await getHooks(this as BaseProgram).RawConfig.promise(config, args);

        Object.assign(config, this.args(args));

        return getHooks(this as BaseProgram).Config.promise(config, args);
    }

    private async post() {
        const stat = stats(this.logger);
        if (stat.error || (this.config.strict && stat.warn)) {
            throw new HandledError('There is some processing errors.');
        }
    }

    private async _action() {
        const args = this.command.optsWithGlobals() as TArgs;
        const config = await this.resolveConfig(args as TArgs);

        // We already parse config in to init method,
        // but there we need to rebuild it,
        // because some extensions may affect config parsing.
        // @ts-ignore
        this['config'] = await this.hookConfig(config, args);

        await this.action(args);
        await this.post();
    }

    private async resolveExtensions(config: Config<BaseConfig>, args: BaseArgs) {
        // args extension paths should be relative to PWD
        const argsExtensions: ExtensionInfo[] = (args.extensions || []).map((ext) => {
            const path = isAbsolute(ext) ? ext : resolve(process.cwd(), ext);
            const options = {};

            return {path, options};
        });

        // config extension paths should be relative to config
        const configExtensions: ExtensionInfo[] = [
            ...this.extensions,
            // ...(config.extensions || []),
        ].map((ext) => {
            const extPath = typeof ext === 'string' ? ext : ext.path;
            const path = isAbsolute(extPath) ? extPath : config.resolve(extPath);
            const options = typeof ext === 'string' ? {} : ext.options || {};

            return {path, options};
        });

        const extensions = [...argsExtensions, ...configExtensions];

        const initialize = async ({
            path,
            options,
        }: {
            path: string;
            options: Record<string, unknown>;
        }) => {
            try {
                const ExtensionModule = require(path);
                const Extension = ExtensionModule.Extension || ExtensionModule.default;

                return new Extension(options);
            } catch (error: any) {
                this.logger.error(`Failed to load extension from ${path}: ${error.message}`);
                throw error;
            }
        };

        const originalRequire = Module.prototype.require;

        // @ts-ignore
        Module.prototype.require = function() {
            const name = arguments[0];
            if (name === '@diplodoc/cli') {
                const realCWD = require.main?.path ? dirname(require.main.path) : process.cwd();
                console.warn('CUSTOM @diplodoc/cli RESOLVED from path', realCWD);
                return originalRequire.apply(this, [realCWD ]);
            }
            // @ts-ignore
            return originalRequire.apply(this, arguments);
        };

        return Promise.all(extensions.map(initialize));
    }
}

function isProgram<TConfig extends BaseConfig, TArgs extends BaseArgs>(
    module: unknown,
): module is BaseProgram<TConfig, TArgs> {
    return Boolean(module && typeof (module as BaseProgram).init === 'function');
}
