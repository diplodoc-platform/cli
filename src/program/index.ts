import type {ICallable, IParent, IProgram} from './types';
import {resolve} from 'node:path';

import {Command, Config, configRoot} from '~/config';
import {YFM_CONFIG_FILENAME} from '~/constants';
import {Build, Publish, Translate} from '~/commands';

import {NAME, USAGE, options} from './config';
import {HandledError, isRelative} from './utils';
import {BaseProgram} from './base';

export type {IProgram, IParent, ICallable};
export {options, HandledError, BaseProgram};

export type ExtensionInfo = {
    path: string;
    options: Record<string, any>;
};

export type ProgramConfig = {
    input: string;
    config: string;
    extensions: ExtensionInfo[];
    quiet: boolean;
    strict: boolean;
};

export type ProgramArgs = {
    input: string;
    config: string;
    extensions: string[];
    quiet: boolean;
    strict: boolean;
};

export class Program
    // eslint-disable-next-line new-cap
    extends BaseProgram<ProgramConfig, ProgramArgs>('Program', {
        config: {
            defaults: () => ({
                extensions: [] as ExtensionInfo[],
            }),
        },
    })
    implements IProgram
{
    readonly command: Command = new Command(NAME)
        .helpOption(true)
        .allowUnknownOption(false)
        .version(
            typeof VERSION !== 'undefined' ? VERSION : '',
            '--version',
            'Output the version number',
        )
        .usage(USAGE);

    readonly build = new Build();

    readonly publish = new Publish();

    readonly translate = new Translate();

    protected options = [
        options.input('./'),
        options.config(YFM_CONFIG_FILENAME),
        options.extensions,
        options.quiet,
        options.strict,
    ];

    private readonly parser: Command = new Command(NAME)
        .addOption(options.input('./'))
        .addOption(options.config(YFM_CONFIG_FILENAME))
        .addOption(options.extensions)
        .helpOption(false)
        .allowUnknownOption(true);

    private readonly modules: ICallable[] = [this.build, this.publish, this.translate];

    async init(argv: string[]) {
        const args = this.parser.parse(argv).opts() as ProgramArgs;
        const config = await this.hookConfig(args);

        this.modules.push(...(await this.resolveExtensions(config, args)));

        for (const module of this.modules) {
            module.apply(this);
        }

        this.apply();
    }

    private async resolveExtensions(config: Config<ProgramConfig>, args: ProgramArgs) {
        // args extension paths should be relative to PWD
        const argsExtensions: ExtensionInfo[] = (args.extensions || []).map((ext: string) => {
            const path = isRelative(ext) ? resolve(ext) : ext;
            const options = {};

            return {path, options};
        });

        // config extension paths should be relative to config
        const configExtensions: ExtensionInfo[] = (config.extensions || []).map(
            (ext: ExtensionInfo | string) => {
                const extPath = typeof ext === 'string' ? ext : ext.path;
                const path = isRelative(extPath) ? resolve(config[configRoot], extPath) : extPath;
                const options = typeof ext === 'string' ? {} : ext.options || {};

                return {path, options};
            },
        );

        const extensions = [...argsExtensions, ...configExtensions];

        const initialize = async ({
            path,
            options,
        }: {
            path: string;
            options: Record<string, unknown>;
        }) => {
            const ExtensionModule = await import(path);
            const Extension = ExtensionModule.Extension || ExtensionModule.default;

            return new Extension(options);
        };

        return Promise.all(extensions.map(initialize));
    }
}
