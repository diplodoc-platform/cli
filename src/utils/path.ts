import {sep} from 'path';
import {Platforms} from '../constants';

export function addSlashPrefix(path: string): string {
    const slashPrefix = path.startsWith(sep) ? '' : sep;

    return `${slashPrefix}${path}`;
}

export function convertBackSlashToSlash(path: string): string {
    if (process.platform === Platforms.WINDOWS) {
        return path.replace(/\\/g, '/');
    }

    return path;
}
