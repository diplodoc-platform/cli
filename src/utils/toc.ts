import {basename, dirname, extname, join} from 'path';

import {YfmToc} from '../models';
import {filterFiles} from '../services/utils';
import {isExternalHref} from './url';
import {getSinglePageAnchorId} from './singlePage';

export function transformToc(toc: YfmToc | null): YfmToc | null {
    if (!toc) {
        return null;
    }

    const localToc: YfmToc = JSON.parse(JSON.stringify(toc));

    if (localToc.items) {
        localToc.items = filterFiles(
            localToc.items,
            'items',
            {},
            {
                removeHiddenTocItems: true,
            },
        );
    }

    const navigationItemQueue = [localToc];

    while (navigationItemQueue.length) {
        const navigationItem = navigationItemQueue.shift();

        if (!navigationItem) {
            continue;
        }

        const {items, href} = navigationItem;

        if (items) {
            navigationItemQueue.push(...navigationItem.items);
        }

        if (href && !isExternalHref(href)) {
            const fileExtension: string = extname(href);
            const filename: string = basename(href, fileExtension);

            navigationItem.href = join(dirname(href), filename);
        }
    }

    return localToc;
}

export function transformTocForSinglePage(
    toc: YfmToc | null,
    options: {root: string; currentPath: string},
) {
    const {root, currentPath} = options;

    if (!toc) {
        return null;
    }

    const localToc: YfmToc = JSON.parse(JSON.stringify(toc));

    if (localToc.items) {
        localToc.items = filterFiles(
            localToc.items,
            'items',
            {},
            {
                removeHiddenTocItems: true,
            },
        );
    }

    function processItems(items: YfmToc[]) {
        items.forEach((item) => {
            if (item.items) {
                processItems(item.items);
            }

            if (item.href && !isExternalHref(item.href)) {
                item.href = getSinglePageAnchorId({root, currentPath, pathname: item.href});
            }
        });
    }

    processItems(localToc.items);

    localToc.singlePage = true;

    return localToc;
}
