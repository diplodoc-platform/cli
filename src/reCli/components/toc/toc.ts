import fs from 'node:fs';
import path from 'node:path';
import {BuildConfig, Run} from '~/commands/build';
import {PresetIndex} from '~/reCli/components/presets/types';
import {TocIndexMap} from '~/reCli/components/toc/types';
import pMap from 'p-map';
import yaml from 'js-yaml';
import {YfmToc} from '~/models';
import {CONCURRENCY} from '~/reCli/constants';
import {getFilePresets} from '~/reCli/components/presets';
import {filterFiles, firstFilterItem, firstFilterTextItems} from '~/services/utils';
import {liquidField} from '~/reCli/components/toc/utils';
import {liquidSnippet} from '@diplodoc/transform/lib/liquid';
import {IncludersError, applyIncluders} from '~/services/includers';
import {logger} from '~/utils';
import {IncludeMode, Stage} from '~/constants';
import shell from 'shelljs';
import {safePath} from '~/reCli/utils';

interface GetTocIndexProps {
    options: BuildConfig;
    presetIndex: PresetIndex;
    run: Run;
    logger: typeof logger;
}

export async function getTocIndex(cwd: AbsolutePath, props: GetTocIndexProps) {
    const {run, options} = props;
    const {ignore} = options;
    const index: TocIndexMap = new Map();

    const tocs = await run.glob('**/toc.yaml', {cwd, ignore});

    const includedTocs: string[] = [];

    await pMap(
        tocs,
        async (tocPath) => {
            const data = await fs.promises.readFile(path.join(cwd, tocPath), 'utf8');
            const toc = yaml.load(data) as YfmToc;
            const parsedToc = await transformToc(toc, {...props, cwd, tocPath});
            if (parsedToc) {
                index.set(tocPath, {toc: parsedToc.toc, copyMap: parsedToc.copyMap});
                includedTocs.push(...parsedToc.includedTocs);
            }
        },
        {concurrency: CONCURRENCY},
    );

    includedTocs.forEach((tocPath) => {
        index.delete(tocPath);
    });

    return index;
}

interface TransformTocProps extends GetTocIndexProps {
    cwd: string;
    tocPath: string;
}

async function transformToc(tocOrig: YfmToc, props: TransformTocProps) {
    const {
        presetIndex,
        tocPath,
        options,
        run: {legacyConfig},
    } = props;
    const {vars, ignoreStage} = options;
    const combinedVars = getFilePresets(presetIndex, vars, tocPath) as Record<string, string>;

    const toc = tocOrig;

    if (toc.stage && ignoreStage.includes(toc.stage)) {
        return null;
    }

    if (toc.title) {
        toc.title = firstFilterTextItems(toc.title, combinedVars, {
            resolveConditions: true,
        });
    }

    if (typeof toc.title === 'string') {
        toc.title = liquidField(toc.title, combinedVars, tocPath, legacyConfig);
    }

    if (typeof toc.navigation === 'string') {
        toc.navigation = liquidField(toc.navigation, combinedVars, tocPath, legacyConfig);
    }

    if (toc.label) {
        toc.label = firstFilterItem(toc.label, combinedVars, {
            resolveConditions: true,
        });
    }

    const copyMap = new Map<string, string>();
    const {includedTocs, items} = await processTocItems(toc.items, combinedVars, {
        ...props,
        copyMap,
    });
    toc.items = items;

    return {toc, includedTocs, copyMap};
}

interface ProcessTocItemsProps extends TransformTocProps {
    copyMap: Map<string, string>;
}

async function processTocItems(
    items: YfmToc[],
    vars: Record<string, string>,
    props: ProcessTocItemsProps,
) {
    const {tocPath, logger, run} = props;
    const {resolveConditions, removeHiddenTocItems} = run.legacyConfig;

    let preparedItems = items;

    /* Should remove all links with false expressions */
    if (resolveConditions || removeHiddenTocItems) {
        try {
            preparedItems = filterFiles(items, 'items', vars, {
                resolveConditions,
                removeHiddenTocItems,
            });
        } catch (err) {
            const error = err as Error;
            logger.error(
                tocPath,
                `Error while filtering toc file: ${tocPath}. Error message: ${error.stack}`,
            );
        }
    }

    /* Should resolve all includes */
    return replaceIncludes(preparedItems, vars, props);
}

interface ReplaceIncludesProps extends TransformTocProps {
    copyMap: Map<string, string>;
}

async function replaceIncludes(
    items: YfmToc[],
    vars: Record<string, string>,
    props: ReplaceIncludesProps,
) {
    const {copyMap, cwd, tocPath, logger} = props;
    const result: YfmToc[] = [];
    const includedTocs: string[] = [];

    for (const item of items) {
        let includedInlineItems: YfmToc[] | null = null;

        if (item.name) {
            item.name = _liquidSubstitutions(item.name, vars, props);
        }

        try {
            await applyIncluders(props.tocPath, item, vars);
        } catch (err) {
            if (err instanceof Error || err instanceof IncludersError) {
                const message = err.toString();

                const file = err instanceof IncludersError ? err.path : props.tocPath;

                logger.error(file, message);
            }
        }

        if (item.include) {
            const {mode = IncludeMode.ROOT_MERGE} = item.include;
            const includeTocPath =
                mode === IncludeMode.ROOT_MERGE
                    ? item.include.path
                    : path.join(path.dirname(tocPath), item.include.path);
            const includeTocDir = path.dirname(includeTocPath);

            try {
                const includeToc = yaml.load(
                    await fs.promises.readFile(
                        path.join(cwd, safePath(includeTocPath)) as AbsolutePath,
                        'utf8',
                    ),
                ) as YfmToc;

                // TODO: use options
                // Should ignore included toc with tech-preview stage.
                if (includeToc.stage === Stage.TECH_PREVIEW) {
                    continue;
                }

                includedTocs.push(includeTocPath);

                if (mode === IncludeMode.MERGE || mode === IncludeMode.ROOT_MERGE) {
                    await copyTocDir(props.run, cwd, includeTocDir, path.dirname(tocPath), copyMap);
                }

                const itemsWithIncluded = (item.items || []).concat(includeToc.items);

                /* Resolve nested toc inclusions */
                const baseTocPath =
                    mode === IncludeMode.LINK ? includeTocPath : path.dirname(tocPath);

                const {items: subItems, includedTocs: subIncludedTocs} = await processTocItems(
                    itemsWithIncluded,
                    vars,
                    {
                        ...props,
                        tocPath: baseTocPath,
                    },
                );

                includedTocs.push(...subIncludedTocs);

                let includedTocItems = subItems;

                /* Make hrefs relative to the main toc */
                if (mode === IncludeMode.LINK) {
                    includedTocItems = _replaceIncludesHrefs(
                        includedTocItems,
                        includeTocDir,
                        path.dirname(tocPath),
                    );
                }

                if (item.name) {
                    item.items = includedTocItems;
                } else {
                    includedInlineItems = includedTocItems;
                }
            } catch (err) {
                const error = err as Error;
                logger.error(
                    tocPath,
                    `Error while including toc: ${includeTocPath} to ${tocPath}. Error: ${error.stack}`,
                );

                continue;
            } finally {
                delete item.include;
            }
        } else if (item.items) {
            const {items: subItems, includedTocs: subIncludedTocs} = await processTocItems(
                item.items,
                vars,
                props,
            );
            item.items = subItems;
            includedTocs.push(...subIncludedTocs);
        }

        if (includedInlineItems) {
            result.push(...includedInlineItems);
        } else {
            result.push(item);
        }
    }

    return {items: result, includedTocs};
}

function _liquidSubstitutions(
    name: string,
    vars: Record<string, string>,
    props: TransformTocProps,
) {
    const {
        tocPath,
        run: {legacyConfig},
    } = props;
    const {outputFormat, applyPresets} = legacyConfig;
    if (outputFormat === 'md' && !applyPresets) {
        return name;
    }

    return liquidSnippet(name, vars, tocPath, {
        conditions: false,
        substitutions: true,
    });
}

function _replaceIncludesHrefs(items: YfmToc[], includeTocDir: string, tocDir: string): YfmToc[] {
    return items.reduce((acc, tocItemOrig) => {
        const tocItem = tocItemOrig;
        if (tocItem.href) {
            tocItem.href = path.relative(tocDir, path.join(includeTocDir, tocItem.href));
        }

        if (tocItem.items) {
            tocItem.items = _replaceIncludesHrefs(tocItem.items, includeTocDir, tocDir);
        }

        if (tocItem.include) {
            const {path: filePath} = tocItem.include;
            tocItem.include.path = path.relative(tocDir, path.join(includeTocDir, filePath));
        }

        return acc.concat(tocItem);
    }, [] as YfmToc[]);
}

async function copyTocDir(
    run: Run,
    cwd: string,
    tocPath: string,
    destDir: string,
    copyMap: Map<string, string>,
) {
    const source = path.join(cwd, safePath(tocPath));
    const target = path.join(cwd, safePath(destDir));

    const files = await run.glob('**/*.*', {
        ignore: ['**/toc.yaml'],
    });

    const dirs = new Set<string>();
    files.forEach((relPath) => {
        const from = path.join(source, relPath);
        const to = path.join(target, relPath);
        const toDir = path.dirname(to);

        if (!dirs.has(toDir)) {
            dirs.add(toDir);
            shell.mkdir('-p', toDir);
        }

        shell.cp(from, to);

        const relSourceFile = path.relative(cwd, from);
        const relTargetFile = path.relative(cwd, to);
        copyMap.set(relSourceFile, relTargetFile);
    });
}
