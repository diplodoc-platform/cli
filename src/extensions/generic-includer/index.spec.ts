import {describe, expect, it} from 'vitest';
import {when} from 'vitest-when';
import {join} from 'node:path';
import {dump} from 'js-yaml';

import {getHooks as getBaseHooks} from '~/core/program';
import {getHooks as getTocHooks} from '~/core/toc';
import {setupBuild, setupRun} from '~/commands/build/__tests__';

import {Extension} from '.';

const prepareExtension = async (globs: [string, RelativePath, NormalizedPath[]][]) => {
    const build = setupBuild();
    const run = setupRun({});
    const extension = new Extension();

    for (const [pattern, cwd, files] of globs) {
        when(run.glob)
            .calledWith(
                pattern,
                expect.objectContaining({
                    cwd: join(run.input, cwd),
                }),
            )
            .thenResolve(files);
    }

    build.apply();
    extension.apply(build);

    await getBaseHooks(build).BeforeAnyRun.promise(run);

    return {build, run, extension};
};

describe('Generic includer', () => {
    it('should work', async () => {
        const {run} = await prepareExtension([
            [
                '**/*.md',
                './test',
                [
                    'index.md',
                    'test.md',
                    'sub/sub-1.md',
                    'sub/sub-2.md',
                    'sub/sub/sub-3.md',
                    'skip/sub/sub-1.md',
                ] as NormalizedPath[],
            ],
        ]);

        const result = await getTocHooks(run.toc)
            .Includer.for('generic')
            .promise(
                {path: 'toc.yaml' as NormalizedPath},
                {
                    input: './test',
                    path: './test/toc.yaml',
                },
            );

        expect(dump(result)).toMatchSnapshot();
    });

    it('should use top path as input root, if input is not specified', async () => {
        const {run} = await prepareExtension([
            [
                '**/*.md',
                './path/test',
                [
                    'index.md',
                    'test.md',
                    'sub/sub-1.md',
                    'sub/sub-2.md',
                    'sub/sub/sub-3.md',
                    'skip/sub/sub-1.md',
                ] as NormalizedPath[],
            ],
        ]);

        const result = await getTocHooks(run.toc)
            .Includer.for('generic')
            .promise(
                {path: 'toc.yaml' as NormalizedPath},
                {
                    path: './path/test/toc.yaml',
                },
            );

        expect(dump(result)).toMatchSnapshot();
    });

    it('should use autotitle option', async () => {
        const {run} = await prepareExtension([
            [
                '**/*.md',
                './test',
                [
                    'index.md',
                    'test.md',
                    'sub/sub-1.md',
                    'sub/sub-2.md',
                    'sub/sub/sub-3.md',
                    'skip/sub/sub-1.md',
                ] as NormalizedPath[],
            ],
        ]);

        const result = await getTocHooks(run.toc)
            .Includer.for('generic')
            .promise(
                {path: 'toc.yaml' as NormalizedPath},
                {
                    input: './test',
                    path: './test/toc.yaml',
                    autotitle: false,
                },
            );

        expect(dump(result)).toMatchSnapshot();
    });
});
