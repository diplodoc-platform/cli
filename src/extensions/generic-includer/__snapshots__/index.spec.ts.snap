// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Generic includer > should use autotitle option 1`] = `
"items:
  - name: index
    href: index.md
  - name: test
    href: test.md
  - name: sub
    items:
      - name: sub-1
        href: sub/sub-1.md
      - name: sub-2
        href: sub/sub-2.md
      - name: sub
        items:
          - name: sub-3
            href: sub/sub/sub-3.md
  - name: skip
    items:
      - name: sub
        items:
          - name: sub-1
            href: skip/sub/sub-1.md
"
`;

exports[`Generic includer > should use top path as input root, if input is not specified 1`] = `
"items:
  - href: index.md
  - href: test.md
  - name: sub
    items:
      - href: sub/sub-1.md
      - href: sub/sub-2.md
      - name: sub
        items:
          - href: sub/sub/sub-3.md
  - name: skip
    items:
      - name: sub
        items:
          - href: skip/sub/sub-1.md
"
`;

exports[`Generic includer > should work 1`] = `
"items:
  - href: index.md
  - href: test.md
  - name: sub
    items:
      - href: sub/sub-1.md
      - href: sub/sub-2.md
      - name: sub
        items:
          - href: sub/sub/sub-3.md
  - name: skip
    items:
      - name: sub
        items:
          - href: skip/sub/sub-1.md
"
`;
